test/unit/test_quilbase.py

```diff
@@ -0,0 +1,64 @@
+from math import pi
+from typing import List
+from pyquil.quil import Program
+from pyquil.quilbase import Gate, MemoryReference, ParameterDesignator
+from pyquil.quilatom import BinaryExp, Qubit, convert_to_rs_qubits
+from pyquil.api._compiler import QPUCompiler
+import pytest
+
+
+@pytest.mark.parametrize(
+    ("name", "params", "qubits"),
+    [
+        ("X", [], [Qubit(0)]),
+        ("CPHASE", [pi / 2], [Qubit(0), Qubit(1)]),
+        ("RZ", [MemoryReference("theta", 0, 1)], [Qubit(0)]),
+        ("RZ", [MemoryReference("alpha", 0) - MemoryReference("beta")], [Qubit(0)]),
+    ],
+    ids=("X-Gate", "CPHASE-Expression", "RZ-MemoryReference", "RZ-MemoryReference-Expression"),
+)
+class TestGate:
```

Comment:
On the note of formatting differences, here's what I've found so far:


1. `pyQuil` evaluates parameter values to check if they contain `pi`. If so, it gets rewritten the parameter as an expression containing `pi`. For example:

```
3.14... => pi
1.57... => pi /2
```

`quil-rs` doesn't do that, instead you'll just get a floating point value. This seems like an important thing to port since `pi` is a Quil constant.

2. `pyQuil` will omit the index of a `MemoryReference` for memory regions of size 1. `quil-rs` will always include an index. For example `MemoryReference("theta", 0, 1)` gets written as `theta` by `pyQuil` and as `theta[0]` by `quil-rs`. In this case, I believe both are valid forms.

3. Some types of `quil-rs` expressions are wrapped in a pair of parantheses. For example where `pyQuil` would generate a binary expression like `alpha[0] - beta[0]` `quil-rs` will output `(alpha[0]-beta[0])`. I think both are valid, but not totally sure if one is preferable to the other.

4. Binary expressions in `quil-rs` have no space between the operator and operands. A stylistic preference.

5. `quil-rs` simplifies any float with a zero-valued decimal to an integer (ie. 0.0 => 0)

6. In Quil, gate definitions can optionally specify a gate type. If omitted, it defaults to `MATRIX`. `pyQuil` will omit the gate type unless it's been explicitly set whereas `quil-rs` will always include a gate type. The former feels more correct, though maybe not a high priority.